ethereum/core/src/integration-test/java/org/hyperledger/besu/ethereum/worldstate/PrunerIntegrationTest.java

```diff
@@ -0,0 +1,244 @@
+    Block parentBlock = blockchain.getChainHeadBlock();
+    for (int i = 0; i < numBlocks; i++) {
+      final MutableWorldState worldState =
+          worldStateArchive.getMutable(parentBlock.getHeader().getStateRoot()).get();
+      gen.createRandomContractAccountsWithNonEmptyStorage(worldState, numAccounts);
+      final Hash stateRoot = worldState.rootHash();
+
+      final Block block =
+          gen.block(
+              BlockOptions.create()
+                  .setStateRoot(stateRoot)
+                  .setBlockNumber(parentBlock.getHeader().getNumber() + 1L)
+                  .setParentHash(parentBlock.getHash()));
+      final List<TransactionReceipt> receipts = gen.receipts(block);
+      blockchain.appendBlock(block, receipts);
+      parentBlock = block;
+    }
+  }
+
+  private Set<BytesValue> collectWorldStateNodes(
+      final Hash stateRootHash, final Set<BytesValue> collector) {
+    final List<Hash> storageRoots = new ArrayList<>();
+    final MerklePatriciaTrie<Bytes32, BytesValue> stateTrie = createStateTrie(stateRootHash);
+
+    // Collect storage roots and code
+    stateTrie
+        .entriesFrom(Bytes32.ZERO, 1000)
+        .forEach(
+            (key, val) -> {
+              final StateTrieAccountValue accountValue =
+                  StateTrieAccountValue.readFrom(RLP.input(val));
+              stateStorage
+                  .get(accountValue.getCodeHash().getArrayUnsafe())
+                  .ifPresent(v -> collector.add(BytesValue.wrap(v)));
+              storageRoots.add(accountValue.getStorageRoot());
+            });
+
+    // Collect state nodes
+    collectTrieNodes(stateTrie, collector);
+    // Collect storage nodes
+    for (Hash storageRoot : storageRoots) {
+      final MerklePatriciaTrie<Bytes32, BytesValue> storageTrie = createStorageTrie(storageRoot);
+      collectTrieNodes(storageTrie, collector);
+    }
+
+    return collector;
+  }
+
+  private void collectTrieNodes(
+      final MerklePatriciaTrie<Bytes32, BytesValue> trie, final Set<BytesValue> collector) {
+    final Bytes32 rootHash = trie.getRootHash();
+    trie.visitAll(
+        (node) -> {
+          if (node.isReferencedByHash() || node.getHash().equals(rootHash)) {
+            collector.add(node.getRlp());
+          }
+        });
+  }
+
+  private MerklePatriciaTrie<Bytes32, BytesValue> createStateTrie(final Bytes32 rootHash) {
+    return new StoredMerklePatriciaTrie<>(
+        worldStateStorage::getAccountStateTrieNode,
+        rootHash,
+        Function.identity(),
+        Function.identity());
+  }
+
+  private MerklePatriciaTrie<Bytes32, BytesValue> createStorageTrie(final Bytes32 rootHash) {
+    return new StoredMerklePatriciaTrie<>(
+        worldStateStorage::getAccountStorageTrieNode,
+        rootHash,
+        Function.identity(),
+        Function.identity());
+  }
+
+  private static class TestInMemoryStorage extends InMemoryKeyValueStorage {
```

Comment:
@mbaxter I was wondering the same. Why did you put this in your mark sweep pruner tests?

Reply:
This was added so that the `InMemoryKeyValueStorage` constructor that accepts an external `Map` parameter does not have to be public (it's protected).  These tests set up spies on this internal storage object.  Would be nice if we could figure out a way to test without needing this.

Reply:
Why is it important to keep it protected? It looks to me like any other dependency injected constructor

Reply:
The general principle is to avoid expanding production API's with functionality that is only useful for testing.  In this particular case, exposing `InMemoryKeyValueStorage`'s internal `Map` kind of breaks encapsulation as well - this object no longer has full control over its internal state.

Reply:
That's fine, a comment is needed though because that's a very subtle reason.